### PR TITLE
arpack: add ISO C binding and top arpack-mpi attribute

### PR DESCRIPTION
**Pull request by fedeinthemix**
I'd like to propose to enable the ISO C bindings which are used by some programs.

In addition, I'd like to add a top attribute `arpack-mpi` with MPI enabled. This makes it easier to use the `arpack` with MPI in flakes.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@ttuegel @dotlambda 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by fedeinthemix** [comment]
ZHF: https://github.com/NixOS/nixpkgs/issues/352882

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -1738,6 +1738,8 @@ with pkgs;
     stdenv = gccStdenv;
   };
 
+  arpack-mpi = callPackage ../by-name/ar/arpack/package.nix { useMpi = true; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:1741]:
```suggestion
  arpack-mpi = arpack.override { useMpi = true; };
```

Comment (fedeinthemix) [reply pkgs/top-level/all-packages.nix:1741]: Good suggestion. I'll push an update commit.

**Comment by fedeinthemix** [comment]
@dotlambda @ttuegel you are indicated as the maintainers of `arpack`. I'd appreciate your review.

**Comment by paparodeo** [comment]
https://hydra.nixos.org/build/282505341

it seems to add single precision api which Accelerate doesn't like so darwin needs to link with openblas.

not quite sure what the difference is before / after but looking at the unit tests run it is now running what looks like a single precision unit tests which segfaults where before this change it wasn't (only 6 tests vs 10).

we've seen this issue with other things using accelerate and single precision on darwin and the fix is simple. will have a PR up shortly.

**Comment by paparodeo** [comment]
pr for darwin https://github.com/NixOS/nixpkgs/pull/367594

**Pull request merged**